src/cpp/src/llm_pipeline.cpp

```diff
@@ -244,60 +233,84 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
-                    auto real_prompt_start = std::find_if(tokenized_prompt.begin(), tokenized_prompt.end(), [&config](int64_t token) { return token != config.eos_token_id; });
-                    tokenized_prompt.erase(tokenized_prompt.begin(), real_prompt_start);
-
-                    sequence_group = std::make_shared<SequenceGroup>(request_id, tokenized_prompt, config, block_size, enable_prefix_caching);
-                    sequence_group->update_processed_tokens_num(tokenized_prompt.size() - 1);
-                }
+        std::vector<SequenceGroup::Ptr> requests;
+        size_t block_size = 1;
+        bool enable_prefix_caching = false;
 
-                sequence_group->set_sequence_group_ptr(sequence_group);
-                requests.push_back(sequence_group);
+        for (size_t request_id = 0; request_id < batch_size; request_id++) {
+            SequenceGroup::Ptr sequence_group;
+            if (is_chat_conversation) {
+                sequence_group = std::make_shared<SequenceGroup>(request_id, m_tokenized_chat_history, config, block_size, enable_prefix_caching);
+                sequence_group->update_processed_tokens_num(m_tokenized_chat_history.get_shape().at(1) - 1);
+            } else {
+                size_t seq_len = input_ids.get_shape().at(1);
+                size_t batch_offset = request_id * seq_len;
+                const int64_t* prompt_start = input_ids.data<const int64_t>() + batch_offset;
+                std::vector<int64_t> tokenized_prompt(prompt_start, prompt_start + seq_len);
+                // in case of multi batch scenario, remove eos_token_id at start of prompt
+                auto real_prompt_start = std::find_if(tokenized_prompt.begin(), tokenized_prompt.end(), [&config](int64_t token) { return token != config.eos_token_id; });
+                tokenized_prompt.erase(tokenized_prompt.begin(), real_prompt_start);
+
+                sequence_group = std::make_shared<SequenceGroup>(request_id, tokenized_prompt, config, block_size, enable_prefix_caching);
+                sequence_group->update_processed_tokens_num(tokenized_prompt.size() - 1);
             }
 
-            Sampler sampler = Sampler(m_tokenizer);
-            std::tie(result, m_selected_beam) = ov::genai::get_lm_encoded_results(m_model_runner, input_ids, concatenated_attention_mask, streamer_ptr,
-                                                                                  sampler, requests, position_ids, std::nullopt, m_selected_beam);
+            sequence_group->set_sequence_group_ptr(sequence_group);
+            requests.push_back(sequence_group);
         }
 
-        if (!is_chat_conversation) {
+        Sampler sampler = Sampler(m_tokenizer);
+        // we can't properly refer to history in case of chat scenario with beam search, so reset_kv_state and use the whole history for each new propmt
+        auto input_tokens = (is_chat_conversation && config.is_beam_search()) ? m_tokenized_chat_history : input_ids;
```

Comment:
I suppose it should `if previous generate was beam_search` ? because previous one can be greedy and we can continue KV cache history